[4.0] Set modified by user to created by user for new items

richard67 opened the pull request:
Pull Request for Issue # .

### Summary of Changes

With the changes for datime database columns recently merged, the handling of modified times was made consistent everywhere so that if something has never been modified, the modified time is equal to the created time (with 1 exception: template overrides, where null for modified time is used to indicate that a component for which an override has been made has meanwhile been uninstalled).

This PR here does the same for the modified by user id, so that at the end if something has never been modified, the modified user id is equal to the created user id.

Together with the first change, the handling is then equal to what filesystems do with files and folder.

Finally, the installer is corrected so it changes the user ID's from hard-coded default to random one also for the workflows.

### Testing Instructions

If you have both MySQL and PostgreSQL, please test on both if possible.

#### Test 1: New installation

1. Have a new installation of the 4.0-dev branch or a 4.0-Alpha-12.
2. Login to backend and switch error reporting to "Maximum" or "Development" in Global Config - Server settings.
3. Create each one new item of following types and don't modify it after having saved it for the first time:

- banner
- category
- article
- contact
- field group
- field
- finder filter
- newsfeed
- tag
- user note
- workflow

4. Check modification times and modified by users in the edit views of the particular items or in database using appropriate tools, and check creator user id and modified by user id for the default workflow.

Result: See section "Actual result" for 4.0 below.

5. Apply the patch of this PR. Attention: It happened to me that the patchtester timed out, so not all changes were applied. To be safe, download the zip of this branch and copy the changed files manually to your J4. If you want to save this work: Download a modified nightly build from the following link which includes this PR [https://test5.richard-fath.de/Joomla_4.0.0-beta1-dev-Development-Full_Package_2019-10-21_pr-26745.zip](https://test5.richard-fath.de/Joomla_4.0.0-beta1-dev-Development-Full_Package_2019-10-21_pr-26745.zip), clear your Joomla folder so it is empty, delete all database tables, unzip the download into the Joomla folder and then perform a new installation. Then skip the next step 6 and continue with 7.
6. Delete file `configuration.php` in your Joomla root folder, and delete all Joomla database tables in PhpMyAdmin or PhpPgAdmin (depending on your database type).
7. Repeat steps 1 to 4 above.

Result: See section "Expected result" below.

#### Test 2: Update of a 3.9.12 or staging.

1. Have a new installation of the staging branch or 3.9.12.
2. Login to backend and switch error reporting to "Maximum" or "Development" in Global Config - Server settings.
3. Create each one new item of following types and don't modify it after having saved it for the first time:

- banner
- category
- article
- contact
- field group
- field
- finder filter
- newsfeed
- tag
- user note

4. Check modification times and modified by users in the edit views of the particular items or in database using appropriate tools.

Result: See section "Actual result" for 3.x below.

5. Update to a patched nightly build which is equal to current 4.0-dev plus the changes in this PR using the "Upload & Update" tab of the Joomla Update component and the update container previously downloaded from following URL: [https://test5.richard-fath.de/Joomla_4.0.0-beta1-dev-Development-Update_Package_2019-10-21_pr-26745.zip](https://test5.richard-fath.de/Joomla_4.0.0-beta1-dev-Development-Update_Package_2019-10-21_pr-26745.zip)
6. Check again modification times and modified by users in the edit views of the particular items or in database using appropriate tools.

Result: See section "Expected result" below.

### Expected result

On 4.0 after patch (new installed and updated): For items which never have been modified, the time of last modification is equal to the time of creation, and the user who has modified it as last is equal to the user who created it.

After new installation, the default workflow has the right creator user id and modified by user id of the super user.

### Actual result

On 4.0 before patch: For items which never have been modified, the time of last modification is equal to the time of creation, but the id of the user who has modified it as last is zero, i.e. the modified by user field is empty in the edit view.

On 3.x or staging before patch: For items which never have been modified, the time of last modification is equal to the (pseudo-)nulldatetime of the database type, and the id of the user who has modified it as last is zero, i.e. the modified by user field is empty in the edit view.

### Documentation Changes Required

None.

richard67 committed:
Set modified user id for new items - SQL

richard67 committed:
Set modified user id for new items - PHP

richard67 committed:
Insert new default data with modified by user = created by user

richard67 committed:
Update DatabaseModel.php

richard67 committed:
Insert default workflow with the right user id

richard67 committed:
Merge branch '4.0-dev' into 4.0-dev-set-modified-by-if-created-null-date

richard67 commented:
@wilsonge Please review. Thanks in advance.

Quy commented on libraries/src/Table/Content.php:
```diff
@@ -354,6 +354,12 @@ public function store($updateNulls = true)
 			}
+
+			// Set modified_by to created_by user if not set
+			if (!$this->modified_by)
```
```suggestion
			if (empty($this->modified_by))
```

richard67 replied:
Thanks.

richard67 committed:
Update libraries/src/Table/Content.php

Co-Authored-By: Quy <quy@fluxbb.org>

richard67 commented:
Please wait with testing, I found some small mistake. For category, modified_user_id is still zero.

richard67 committed:
Update ContactTable.php

richard67 committed:
Modified times check conistent with existing times checks

richard67 committed:
Update ContactTable.php

richard67 commented:
PR is ready for test.

Important hint for testers: It happened to me that the (old) patchtester timed out, so not all changes have been applied. So maybe it is better to download the zip of the complete branch here and copy the files manually into your testing environment for the 1st test with new installation of J4.

richard67 committed:
Merge branch '4.0-dev' into 4.0-dev-set-modified-by-if-created-null-date

richard67 committed:
Merge branch '4.0-dev' into 4.0-dev-set-modified-by-if-created-null-date

Quy commented on administrator/components/com_banners/Table/BannerTable.php:
```diff
@@ -141,6 +141,12 @@ public function check()
 		}
 
+		// Set modified_by to created_by if not set
+		if (!$this->modified_by)
```
```suggestion
		if (empty($this->modified_by))
```

richard67 committed:
Update administrator/components/com_banners/Table/BannerTable.php

Co-Authored-By: Quy <quy@fluxbb.org>

richard67 committed:
Merge branch '4.0-dev' into 4.0-dev-set-modified-by-if-created-null-date

richard67 committed:
Merge branch '4.0-dev' into 4.0-dev-set-modified-by-if-created-null-date

brianteeman commented:
Sorry I am confused here. If an item has never been modified why are we setting a value. It seems wrong to me.

richard67 commented:
@brianteeman Was agreed with George to do it this way. For the modified times and user ids it has already been done at several places in PHP in the table classes in check and/or store functions. For the modified times we then have done this with the recently merged PRs for the nulldates, but to do the same with the modified user IDs has been forgotten. This PR here just adds that. At the end behavior is consistent everywhere and is same as filesystems handle modification times and users for files and folders.

I know this will not convince you, it shall only be an explanation for the reasons behind it.

richard67 commented:
@brianteeman P.S.: And you are not alone with your opinion, Sharky or Harald would agree with you, I think, and surely more people. I did not want to decide on that and would have done it the one or the other way and was ok with both and so left the decision to George at the end.

SharayuYadav commented:
I have tested this item :red_circle: unsuccessfully on dffa542ca0c73701e1e0c7ede98d6cf416abc393<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26745">issues.joomla.org/tracker/joomla-cms/26745</a>.</sub>

SharayuYadav commented:
I have tested this item :red_circle: unsuccessfully on dffa542ca0c73701e1e0c7ede98d6cf416abc393<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26745">issues.joomla.org/tracker/joomla-cms/26745</a>.</sub>

ttplpoojak commented:
I have tested this item :red_circle: unsuccessfully on dffa542ca0c73701e1e0c7ede98d6cf416abc393<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26745">issues.joomla.org/tracker/joomla-cms/26745</a>.</sub>

wilsonge commented on installation/sql/postgresql/joomla.sql:
```diff
@@ -270,12 +270,12 @@ COMMENT ON COLUMN "#__categories"."metadata" IS 'JSON encoded metadata propertie
-(4, 29, 1, 5, 6, 1, 'uncategorised', 'com_contact', 'Uncategorised', 'uncategorised', '', '', 1, 0, NULL, 1, '{"category_layout":"","image":""}', '', '', '{"author":"","robots":""}', 42, CURRENT_TIMESTAMP, 0, CURRENT_TIMESTAMP, 0, '*', 1),
-(5, 30, 1, 7, 8, 1, 'uncategorised', 'com_newsfeeds', 'Uncategorised', 'uncategorised', '', '', 1, 0, NULL, 1, '{"category_layout":"","image":""}', '', '', '{"author":"","robots":""}', 42, CURRENT_TIMESTAMP, 0, CURRENT_TIMESTAMP, 0, '*', 1),
-(7, 32, 1, 9, 10, 1, 'uncategorised', 'com_users', 'Uncategorised', 'uncategorised', '', '', 1, 0, NULL, 1, '{"category_layout":"","image":""}', '', '', '{"author":"","robots":""}', 42, CURRENT_TIMESTAMP, 0, CURRENT_TIMESTAMP, 0, '*', 1);
+(1, 0, 0, 0, 11, 0, '', 'system', 'ROOT', 'root', '', '', 1, 0, NULL, 1, '{}', '', '', '{}', 42, CURRENT_TIMESTAMP, 42, CURRENT_TIMESTAMP, 0, '*', 1),
```
This looks wrong. The `updateUserIds` in the installer is replacing these with the random user ID when the ID is 0 - why are we making this hardcoded 42 which isn't the super user ID?

richard67 replied:
@wilsonge It is done like that at several other places with 42 in joomla.sql where we insert data having user ID's, and those hard coded user ID's are later replaced in routine "updateUserIds" in file "installation/src/Model/DatabaseModel.php". But if you want I change the 42 back to 0 here like it is right now without this PR.

richard67 replied:
Ok I change to 0 here.

richard67 replied:
@wilsonge Fuck, I changed it and now I see I was right with not using 0, see here: [https://github.com/joomla/joomla-cms/blob/0d79b5de7f39a4e58b166316235a25fe39d86d6b/installation/src/Model/DatabaseModel.php#L848](https://github.com/joomla/joomla-cms/blob/0d79b5de7f39a4e58b166316235a25fe39d86d6b/installation/src/Model/DatabaseModel.php#L848). The replacement is not done if the user id is zero. And look on the diff now after my (wrong) last change: It shows that for mysql it was 42 before this PR for categories and tags table. Please check and let me know if I shall change back to 42.

richard67 replied:
I've rolled back to 42 now. Was like this before this PR, and the above link shows 0 would not work. If the 42 is an issue it may be solved with a future PR.

wilsonge replied:
🤦‍♂ sorry yup it was right

richard67 commented:
@SharayuYadav @ttplpoojak Sure you've tested as described in the testing instructions, i.e. after having applied the patch with patchtester, removed configuration.php and deleted database tanles and then made a new installation? If so, please report back at which step what went wrong. Otherwise, if you have tested in the wrong way, please change back your testing result in the issue tracker to "I have not tested this item", and if time test again in the right way.

richard67 committed:
Change user if 42 to 0 for new inserted data

richard67 committed:
Undo changes in files 4.0.0-2018-05-15.sql

richard67 committed:
Revert "Change user if 42 to 0 for new inserted data"

This reverts commit 69d45a35dbd67ddaa3b82f76d350c080ef5cda33.

richard67 committed:
Merge branch '4.0-dev' into 4.0-dev-set-modified-by-if-created-null-date

wilsonge commented:
Thanks!

richard67 commented:
Thanks too.
